core/trino-main/src/main/java/io/trino/operator/aggregation/partial/ActivePartialAggregationControl.java

```diff
@@ -0,0 +1,109 @@
+ * limitations under the License.
+ */
+package io.trino.operator.aggregation.partial;
+
+import io.trino.operator.HashAggregationOperator;
+
+/**
+ * Controls whenever partial aggregation is enabled across all {@link HashAggregationOperator}s
+ * for a particular plan node on a single node.
+ * Partial aggregation is disabled once enough rows has been processed ({@link #minNumberOfRowsProcessed})
+ * and the ratio between output(unique) and input rows is too high (> {@link #uniqueRowsRatio}).
+ * <p>
+ * The class is thread safe and objects of this class are used potentially by multiple threads/drivers simultaneously.
+ * Different threads either:
+ * - modify fields via synchronized {@link #onPartialAggregationFlush}.
+ * - read volatile {@link #partialAggregationDisabled} (volatile here gives visibility).
+ */
+public class ActivePartialAggregationControl
+        implements PartialAggregationControl
+{
+    private final long minNumberOfRowsProcessed;
+    private final double uniqueRowsRatio;
+
+    private volatile boolean partialAggregationDisabled;
+    private long totalRowProcessed;
+    private long totalUniqueRowsProduced;
+
+    public ActivePartialAggregationControl(long minNumberOfRowsProcessedToDisable, double uniqueRowsRatio)
+    {
+        this.minNumberOfRowsProcessed = minNumberOfRowsProcessedToDisable;
+        this.uniqueRowsRatio = uniqueRowsRatio;
+    }
+
+    public synchronized void onPartialAggregationFlush(long rowsProcessed, long uniqueRowsProduced)
+    {
+        totalRowProcessed += rowsProcessed;
+        totalUniqueRowsProduced += uniqueRowsProduced;
+        if (shouldDisablePartialAggregation()) {
+            partialAggregationDisabled = true;
+        }
+    }
+
+    private boolean shouldDisablePartialAggregation()
+    {
+        return totalRowProcessed >= minNumberOfRowsProcessed
+                && ((double) totalUniqueRowsProduced / totalRowProcessed) > uniqueRowsRatio;
+    }
+
+    @Override
+    public PartialAggregationTracker newTracker()
+    {
+        return new ActivePartialAggregationTracker();
+    }
+
+    /**
+     * Tracks number of input rows processed and unique groups produced
+     * per single Partial Aggregation hash table (i.e. per flush).
+     */
+    public class ActivePartialAggregationTracker
+            implements PartialAggregationTracker
+    {
+        private long numberOfRowsProcessed;
+        private long numberOfUniqueRowsProduced;
+
+        @Override
+        public boolean isPartialAggregationDisabled()
+        {
+            return partialAggregationDisabled;
+        }
+
+        @Override
+        public void addRowsProcessed(int positionCount)
+        {
+            numberOfRowsProcessed += positionCount;
+        }
+
+        @Override
+        public void addUniqueRowsProduced(int positionCount)
+        {
+            numberOfUniqueRowsProduced += positionCount;
+        }
+
+        @Override
+        public void onFlush()
```

Comment:
you can just make it `onFlush(int totalPositionCount, int uniquePositionCount)`. Make row tracking internal to `HashAggregationOperator`. Then you don't even need this tracker at all

Reply:
But you call these methods anyway. What we do here is simple row tracking. We don't need factory, interfaces and tracker for that. It's an overkill to just do simple increments